src/test/java/reactor/pool/CommonPoolTest.java

```diff
@@ -2365,4 +2366,82 @@ void raceShutdownAndPreAcquiredReleaseWithEviction(Function<PoolBuilder<AtomicIn
 		assertThat(ai).as("creates and destroys stabilizes to 0").hasValue(0);
 	}
 
+	@ParameterizedTest
+	@MethodSource("allPools")
+	void maxPendingZero(Function<PoolBuilder<Integer, ?>, AbstractPool<Integer>> configAdjuster) {
+		AtomicInteger source = new AtomicInteger();
+		PoolBuilder<Integer, PoolConfig<Integer>> configBuilder = PoolBuilder
+				.from(Mono.fromSupplier(source::incrementAndGet))
+				.sizeBetween(1, 2)
+				.maxPendingAcquire(0);
+
+		AbstractPool<Integer> pool = configAdjuster.apply(configBuilder);
+
+		assertThat(pool.warmup().block(Duration.ofSeconds(1))).as("warmup").isOne();
+
+		//there is one idle resource
+		assertThat(pool.acquire().block(Duration.ofSeconds(1)))
+				.as("acquire on idle")
+				.isNotNull()
+				.hasFieldOrPropertyWithValue("poolable", 1);
+
+		//there is now idle resource, but still capacity
+		assertThat(pool.acquire().block(Duration.ofSeconds(1)))
+				.as("acquire on allocate")
+				.isNotNull()
+				.hasFieldOrPropertyWithValue("poolable", 2);
+
+		//there is now idle resource, but still capacity
+		assertThatExceptionOfType(PoolAcquirePendingLimitException.class)
+				.isThrownBy(() -> pool.acquire().block(Duration.ofSeconds(1)))
+				.as("acquire on maxPending")
+				.withMessage("No pending allowed and pool has reached allocation limit");
+	}
+
+	@ParameterizedTest
+	@MethodSource("allPools")
+	void maxPendingOne(Function<PoolBuilder<Integer, ?>, AbstractPool<Integer>> configAdjuster)
+			throws InterruptedException {
+		AtomicInteger source = new AtomicInteger();
+		PoolBuilder<Integer, PoolConfig<Integer>> configBuilder = PoolBuilder
+				.from(Mono.fromSupplier(source::incrementAndGet))
+				.sizeBetween(1, 2)
+				.maxPendingAcquire(1);
+
+		AbstractPool<Integer> pool = configAdjuster.apply(configBuilder);
+
+		assertThat(pool.warmup().block(Duration.ofSeconds(1))).as("warmup").isOne();
+
+		//there is one idle resource
+		assertThat(pool.acquire().block(Duration.ofSeconds(1)))
+				.as("acquire on idle")
+				.isNotNull()
+				.hasFieldOrPropertyWithValue("poolable", 1);
+
+		//there is now idle resource, but still capacity
+		assertThat(pool.acquire().block(Duration.ofSeconds(1)))
+				.as("acquire on allocate")
+				.isNotNull()
+				.hasFieldOrPropertyWithValue("poolable", 2);
+
+		//there is now idle resource, no capacity, but pending 1 is allowed
+		//in order to test the scenario where we've reached maxPending, this pending must not be cancelled
+		final CountDownLatch latch = new CountDownLatch(1);
```

Comment:
nit: alternatively, can be `pool.acquire().timeout(Duration.ofSeconds(1)).toFuture()`, to avoid `latch` + `endingSignal`